Add option to install modules concurrently

oesgalha opened the pull request:
This pull request add the option to install/update modules concurrently.

It adds a `pool_size` setting with a default value equals to `1`:

```yaml
# /etc/puppetlabs/r10k/r10k.yaml
pool_size: 5
```

When the `Puppefile#accept` method is called, if the `pool_size` value is 1, the regular sequential install/update process starts. If the `pool_size` is greater than 1 a thread pool with `pool_size` threads is created to fetch the modules. A `pool_size` with value less or equal to zero is not allowed.

Since the default setting value is 1 this feature is _opt-in_.

The install/update process is heavily I/O based so there are some visible performance gains in the total (wall) time to install modules declared in a Puppetfile.

We've been using this patch for almost 5 months in production with no reported problems.

## Benchmarks

Puppetfile with 9 modules

|Flavor|Time|Comparison|
|--------|-------|----------------|
|Vanilla|69.846 s|base|
|pool_size=5|19.985  s|-71.39%|

Puppetfile with 608 modules

|Flavor|Time|Comparison|
|--------|-------|----------------|
|Vanilla|8.705 m|base|
|pool_size=5|2.056  m|-76.38%|
|pool_size=10|1.901  m|-78.16%|

oesgalha committed:
Implements a thread pool to update modules concurrently

Add a pool_size setting to r10k.yaml and uses 1 as a default value.
The feature is opt-in, if the r10k.yaml does not specify a number greater than 1
the serial update is used (same behaviour before this commit).

oesgalha committed:
Add support to non-duplicable values in r10k.yaml

Integers can't be dupped until ruby 2.4, in order to support older ruby versions
and integers in the config file, the parent container fall back strategy
should directly assign when the value doesn't support dup

oesgalha committed:
Add specs for the thread based module install

Add specs for the `pool_size` setting and the Puppetfile `accept`
execution when the `pool_size` setting is greater than one.

Sharpie commented on lib/r10k/puppetfile.rb:
```diff
@@ -146,14 +151,48 @@ def purge_exclusions
+    Thread.new do
+      begin
+        while mod = mods_queue.pop(true) do mod.accept(visitor) end
+      rescue ThreadError
```
What sorts of errors could trigger this rescue? Should they be logged?

reidmv committed:
Log error message on rescued ThreadError

In the event something goes wrong and we need to terminate a thread, we
should make that occurance visible.

Sharpie reviewed (approved):
Looks good 👍
